plugins/modules/ec2_vpc_vgw.py

```diff
@@ -198,39 +212,50 @@ def get_vgw_info(vgws):
         return vgw_info
 
 
-def wait_for_status(client, module, vpn_gateway_id, status):
+def wait_for_status(
+    client, module: AnsibleAWSModule, vpn_gateway_id: Union[str, List[str]], desired_status: str
+) -> Tuple[bool, Any]:
     polling_increment_secs = 15
     max_retries = module.params.get("wait_timeout") // polling_increment_secs
-    status_achieved = False
+    try:
+        wait_for_resource_state(client, module, "vpn_gateway_exists", VpnGatewayIds=vpn_gateway_id)
+        if desired_status == "attached":
+            wait_for_resource_state(
+                client,
+                module,
+                "vpn_gateway_attached",
+                VpnGatewayIds=vpn_gateway_id,
+                delay=polling_increment_secs,
+                max_attempts=max_retries,
+            )
+        elif desired_status == "detached":
+            wait_for_resource_state(
+                client,
+                module,
+                "vpn_gateway_detached",
+                VpnGatewayIds=vpn_gateway_id,
+                delay=polling_increment_secs,
+                max_attempts=max_retries,
+            )
+        else:
+            module.fail_json(msg=f"Unsupported status: {desired_status}")
 
-    for x in range(0, max_retries):
-        try:
-            response = find_vgw(client, module, vpn_gateway_id)
-            if response[0]["VpcAttachments"][0]["State"] == status:
-                status_achieved = True
-                break
-            else:
-                time.sleep(polling_increment_secs)
-        except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
-            module.fail_json_aws(e, msg="Failure while waiting for status update")
+        response = find_vgw(client, module, vpn_gateway_id)
+        status_achieved = response[0]["VpcAttachments"][0]["State"] == desired_status
 
-    result = response
-    return status_achieved, result
+    except AnsibleEC2Error as e:
+        module.fail_json_aws(e)
 
+    return status_achieved, response
 
-def attach_vgw(client, module, vpn_gateway_id):
-    params = dict()
-    params["VpcId"] = module.params.get("vpc_id")
+
+def attach_vgw(client, module: AnsibleAWSModule, vpn_gateway_id: str) -> Any:
```

Comment:
Is check_mode handled for this operation?

Reply:
fixed